src/detect.rs

```diff
@@ -18,3 +23,221 @@ pub enum DetectOutcome {
     Pass(BuildPlan),
     Fail,
 }
+
+pub struct DetectContextBuilder<P: Platform, BM: DeserializeOwned> {
+    app_dir: Option<PathBuf>,
+    buildpack_dir: Option<PathBuf>,
+    stack_id: String,
+    platform: Option<P>,
+    platform_dir: Option<PathBuf>,
+    buildpack_toml: Option<BuildpackToml<BM>>,
+    buildpack_toml_path: Option<PathBuf>,
+    buildpack_toml_string: Option<String>,
+}
+impl<P: Platform, BM: DeserializeOwned> DetectContextBuilder<P, BM> {
+    pub fn new(stack_id: impl AsRef<str>) -> Self {
+        DetectContextBuilder {
+            app_dir: None,
+            buildpack_dir: None,
+            buildpack_toml_path: None,
+            buildpack_toml_string: None,
+            platform_dir: None,
+            stack_id: String::from(stack_id.as_ref()),
+            platform: None,
+            buildpack_toml: None,
+        }
+    }
+
+    /// Used for testing purposes to set a minimal viable context
+    ///
+    ///```
+    /// use tempfile;
+    /// use libcnb::DetectContextBuilder;
+    ///
+    /// pub fn detect(context: libcnb::GenericDetectContext) {}
+    ///
+    /// let temp_dir = tempfile::tempdir().unwrap();
+    /// let context = DetectContextBuilder::new("heroku-20")
+    ///     .buildpack_toml_from_string(include_str!("../examples/example-02-ruby-sample/buildpack.toml"))
+    ///     .from_temp_dir(&temp_dir)
+    ///     .build::<std::io::Error>()
+    ///     .unwrap();
+    ///
+    /// detect(context);
+    ///```
+    pub fn from_temp_dir(mut self, temp_dir: &tempfile::TempDir) -> DetectContextBuilder<P, BM> {
```

Comment:
```suggestion
    pub fn from_temp_dir(mut self, temp_dir: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
```

Should we use `Path` here? We don't do anything special with `TempDir` here.

Reply:
I wanted to restrict it to make sure people were using this in a way I expected. I'm worried someone might point it to a their current directory and suddenly new "app" and "buildpack" directories are created and modified.

Reply:
The downside is now we're requiring this dev dependency in production builds. I feel like forcing the tempdir doesn't buy us very much. I expect people will read the docs and use it as intended.

Reply:
>  I expect people will read the docs

Bold claim 😹

> The downside is now we're requiring this dev dependency in production builds. 

That's a good point. It also forces a specific tempdir library, which isn't actually needed. I'll switch to using a Path instead.

Reply:
I refactored all this logic, it no longer takes a path. Instead it generate and retains it's own tempdir.